pmlc/rt/level_zero/level_zero_invocation.h

```diff
@@ -0,0 +1,151 @@
+                 std::string name);
+
+  /// Returns LevelZero event object.
+  ze_event_handle_t getEvent() const { return event; }
+  /// Returns kind of operation that this event describes.
+  LevelZeroActionKind getKind() const { return kind; }
+  /// Returns name of operation that this event describes.
+  const std::string &getName() const { return name; }
+
+  /// Blocks execution until all `events` have finished executing.
+  static void wait(const std::vector<LevelZeroEvent *> &events);
+
+private:
+  ze_event_handle_t event;
+  LevelZeroActionKind kind;
+  std::string name;
+};
+
+// LevelZeroInvocation encapsulates a particular run of a network on a LevelZero
+// device. It's instantiated and managed from the JITted network code, using
+// callbacks in level_zero_wrappers.cc.
+class LevelZeroInvocation {
+public:
+  explicit LevelZeroInvocation(LevelZeroDevice *device);
+  ~LevelZeroInvocation();
+
+  /// Allocates memory on LevelZero device with specified size in bytes.
+  LevelZeroMemory *allocateMemory(size_t bytes);
+  /// Releases memory obtained from `allocateMemory` call.
+  /// But as kernel may not really run here, may just do record.
+  void deallocateMemory(LevelZeroMemory *memory);
+  /// Enqueues operation to read data from `src` memory into `dst` pointer.
+  /// Read will start executing only after operations connected to `deps`
+  /// events are finished.
+  /// Returns event describing status of this operation.
+  LevelZeroEvent *enqueueRead(LevelZeroMemory *src, void *dst,
+                              const std::vector<LevelZeroEvent *> &deps);
+  /// Enqueues operation to write data from `src` pointer into `dst` memory.
+  /// Write will start executing only after operations connected to `deps`
+  /// events are finished.
+  /// Returns event describing status of this operation.
+  LevelZeroEvent *enqueueWrite(LevelZeroMemory *dst, void *src,
+                               const std::vector<LevelZeroEvent *> &deps);
+  /// Creates LevelZero kernel from SPIRV binary.
+  LevelZeroKernel *createKernelFromIL(char *data, size_t bytes,
+                                      const char *name);
+  /// Enqueues kernel execution on device with specified global and local work
+  /// sizes. Kernel object needs to be destroyed manually. Returns
+  /// event describing status of kernel execution.
+  LevelZeroEvent *enqueueKernel(LevelZeroKernel *kernel, ze_group_count_t gws,
+                                ze_group_count_t lws);
+  /// Not used now
+  LevelZeroEvent *enqueueBarrier(const std::vector<LevelZeroEvent *> &deps);
+  /// Forces starting execution of previosuly enqueued operations that aren't
+  /// blocked by any dependencies.
+  void flush();
+  /// Blocks untill all previously enqueued operations are finished.
+  void finish();
+
+private:
+  std::shared_ptr<LevelZeroDevice> device;
+  LevelZeroQueueUser queueUser;
+  lzu::zeEventPool eventPool;
+  // clear list
```

Comment:
What does this comment mean?

Reply:
Ok, these are resources needs to be released. Maybe reword it to something more clear.

Reply:
Done